### PR TITLE
Force semantic versioning and release automatically

### PR DESCRIPTION
**Pull request by fpagnoux**
The goal of this PR is to move toward semantic versioning and continuous releasing. What it does: 

* Force version number incrementation through CI
* Force changelog editing through CI
* Publish tag after merging
* Publish on pypi after tagging

Todo: 
- [x] Document semantic versioning in our context in the doc or wiki
- [x] Update the releasing process in the doc (internationalisation is not part of the release process anymore)

### CODE REVIEW

File: release-tag.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+version=`python setup.py --version`
+eval "$(ssh-agent -s)" #start the ssh agent
+chmod 400 ./openfisca_bot
```
Comment (fpagnoux) [inline release-tag.sh]: Without doing this, ssh-add will not add the key as it is not safe.
Is there a security problem with, during the build, decrypting the ssh key into a clear text file before adding it ?
The only risk I would see is that someone could, by mistake, print the ssh key - or the travis key that decrypts it - on the log by adding a wrong command to the release script or the travis configuration file. It doesn't sound really likely.

An alternative would be to try to use a github token, but anyway, if we want to do real continuous deployment in the future, we will need a ssh key.

@MattiSG ?

Comment (MattiSG) [inline release-tag.sh]: Yes, it is a very strong security risk if you build PRs from forks. As an attacker, I just have to open a PR with a modified release script that prints the contents of the file.

You should rely on the security mechanisms from your CI engine. For example, CircleCI [disables many variables](https://circleci.com/docs/fork-pr-builds/) when building PRs from forks. If you add the SSH key through their web UI, everything will be handled for you.

I haven't looked up the equivalent with Travis, but I guess they have something too.

File: run-travis-tests.sh [hunk 2]
```diff
@@ -7,6 +7,24 @@
...
+	then
+		set +x
+		echo "Version $version already exists. Please update version number in setup.py before merging this branch into master."
+		exit 1
+	fi
+
+	if git diff-index master --quiet CHANGELOG.md
+	then
+		set +x
+		echo "CHANGELOG.md has not been modified. Please update it before merging this branch into master."
```
Comment (MattiSG) [inline run-travis-tests.sh]: 👍

File: release-tag.sh (hunk 1 above)
Comment (fpagnoux) [inline release-tag.sh]: I don't think travis let you add an ssh key to the project settings like Circle CI 😢 .

However, encrypted files are [not available for pull requests coming from forks](https://docs.travis-ci.com/user/pull-requests#Security-Restrictions-when-testing-Pull-Requests), so I don't think the kind of attack you describe is possible.

There is still the risk of a mistake from a trusted developper though. Do you think this risk is enough to spend some time migrating to Circle-CI ?

Comment (MattiSG) [inline release-tag.sh]: No. Disabling encryption for incoming PRs is fine  :)

File: .travis.yml [hunk 3]
```diff
@@ -13,3 +13,18 @@ script: bash run-travis-tests.sh
 sudo: false
 virtualenv:
   system_site_packages: true
+before_deploy:
+- openssl aes-256-cbc -K $encrypted_f6aff9134f80_key -iv $encrypted_f6aff9134f80_iv -in openfisca_bot.enc -out openfisca_bot -d
+deploy:
+  - provider: script
+    skip_cleanup: true
+    script: ./release.sh
```
Comment (MattiSG) [inline .travis.yml]: That script doesn't exist. It is named `release-tag.sh`.

File: release-tag.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+ssh-add ./openfisca_bot
```
Comment (MattiSG) [inline release-tag.sh]: There is a dependency on this file being there, and thus on Travis having decrypted the file. This should at least be documented at the top of the file.

**Comment by MattiSG** [comment]
Please test all that on a separate, temporary branch that will have its target branch for deployment as itself, so that you can see whether the whole thing works.

> (testing CI adjustments is one of the most annoying things to do)

File: .travis.yml (hunk 3 above)
Comment (fpagnoux) [inline .travis.yml]: 👍

**Comment by fpagnoux** [comment]
It is hard to test, as the tag pushing will only happen when merging on master...
I can try a slightly different version that push tags when merging to a test integration branch.
I tested most of it [here](https://github.com/fpagnoux/test-travis) though but yea, we're not protected from an error in encryption, or something like that.

**Comment by MattiSG** [comment]
> will only happen when merging on master...

thus:

> temporary branch that will have its target branch for deployment as itself,

Even beyond encryption, my last comment on the name of the script is the kind of error you don't see otherwise  😞

**Comment by fpagnoux** [comment]
Ok I just understood the meaning of this sentence ;)

**Comment by fpagnoux** [comment]
It works: see those [two](https://travis-ci.org/openfisca/openfisca-france/builds/126095718) [builds](https://travis-ci.org/openfisca/openfisca-france/builds/126095718) for [this branch](https://github.com/openfisca/openfisca-france/compare/releasing...CI-testing-integration).

**Comment by MattiSG** [comment]
Looking at the build, the following errors appear:

```
Installing deploy dependencies
sudo: must be setuid root
```

(due to container-based Travis build)

- - - - -

```
git checkout 1.1.dev4
error: pathspec '1.1.dev4' did not match any file(s) known to git.
```

I don't understand that one: I don't see in your script where you `git checkout`.

- - - - 

At the end of publishing:

```
InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
```

- - - - -

At the end of publishing:

```
running upload_docs
creating build/docs
error: no files found in upload directory 'build/docs'
```

**Comment by fpagnoux** [comment]
>```
git checkout 1.1.dev4
error: pathspec '1.1.dev4' did not match any file(s) known to git.
```

>I don't understand that one: I don't see in your script where you `git checkout`.

[Here](https://github.com/openfisca/openfisca-france/blob/master/run-travis-tests.sh#L14). It existed before this PR.

**Comment by fpagnoux** [comment]
About the 3 others warnings, they happen when publishing a version with the built-in pypi deployment tool, and we have little control about how it's done.
The `'build/docs` for instance is just due to the fact that travis [doesn't allow you to deactivate docs uploading](https://github.com/travis-ci/dpl/issues/334).
These warnings are not blocking for publishing.

**Comment by fpagnoux** [comment]
Seems GTM to me. Any last thoughts ?
@cbenz @laem

**Comment by fpagnoux** [comment]
See https://github.com/openfisca/openfisca-gitbook/pull/11 for doc.

**Pull request merged**